diff -uNr a/drivers/nvmem/Makefile b/drivers/nvmem/Makefile
--- a/drivers/nvmem/Makefile	2023-02-25 02:50:37.520000000 +0800
+++ b/drivers/nvmem/Makefile	2023-02-25 02:53:27.176000000 +0800
@@ -55,3 +55,5 @@
 nvmem_zynqmp_nvmem-y		:= zynqmp_nvmem.o
 obj-$(CONFIG_SPRD_EFUSE)	+= nvmem_sprd_efuse.o
 nvmem_sprd_efuse-y		:= sprd-efuse.o
+obj-$(CONFIG_RTK_EFUSE)         += nvmem_rtk-efuse.o
+nvmem_rtk-efuse-y		:= rtk-efuse.o
\ 文件尾没有换行符
