diff -uNr a/drivers/usb/dwc3/core.h b/drivers/usb/dwc3/core.h
--- a/drivers/usb/dwc3/core.h	2023-02-28 21:44:23.792000000 +0800
+++ b/drivers/usb/dwc3/core.h	2023-02-28 21:45:05.764000000 +0800
@@ -554,6 +554,10 @@
 #define DWC3_DEV_IMOD_INTERVAL_SHIFT	0
 #define DWC3_DEV_IMOD_INTERVAL_MASK	(0xffff << 0)
 
+#if 1 // USB_PATCH_BY_RTK
+#define DWC3_DEVICE_IMODI(n)		((0xffff & (n)))
+#endif // USB_PATCH_BY_RTK
+
 /* OTG Configuration Register */
 #define DWC3_OCFG_DISPWRCUTTOFF		BIT(5)
 #define DWC3_OCFG_HIBDISMASK		BIT(4)
