Index: gcompat-1.1.0/Makefile
===================================================================
--- gcompat-1.1.0.orig/Makefile
+++ gcompat-1.1.0/Makefile
@@ -76,7 +76,7 @@ ${LIBGCOMPAT_NAME}: ${LIBGCOMPAT_OBJ}
 ${LIBGCOMPAT_OBJ}: ${LIBGCOMPAT_INCLUDE}
 
 ${LOADER_NAME}: ${LOADER_OBJ}
-	${CC} ${CFLAGS} ${LDFLAGS} -static-pie -o ${LOADER_NAME} ${LOADER_OBJ}
+	${CC} ${CFLAGS} ${LDFLAGS} -static-pie -static -pie -z text -o ${LOADER_NAME} ${LOADER_OBJ}
 
 .c.o:
 	${CC} ${CPPFLAGS} ${CFLAGS} -c -D_BSD_SOURCE \
